x/ibc/03-connection/types/version.go

```diff
@@ -1,39 +1,120 @@
+	version := versionNumber
+	for i, feature := range featureSet {
+		if i == 0 {
+			version = fmt.Sprintf("%s-%s", version, feature)
+		} else {
+			version = fmt.Sprintf("%s,%s", version, feature)
+		}
+	}
+
+	return version
+}
+
+// GetVersionNumber returns the version number of a given connection version.
+func GetVersionNumber(version string) (string, error) {
+	splitVersion := strings.Split(version, "-")
+
+	// check if no dash exists in the version
+	if len(splitVersion) < 2 {
+		return "", sdkerrors.Wrapf(
+			ErrInvalidVersion,
+			"could not retrieve the version number for version (%s)", version,
+		)
+	}
+
+	return strings.TrimSpace(splitVersion[0]), nil
+}
+
+// GetFeatureSet returns the feature set for a given connection version.
+func GetFeatureSet(version string) ([]string, error) {
+	// only split version number and feature set
+	splitVersion := strings.SplitN(version, "-", 1)
+	if len(splitVersion) < 2 {
+		return nil, sdkerrors.Wrapf(
+			ErrInvalidVersion,
+			"failed to retrieve the feature set for version (%s)", version,
+		)
+	}
+
+	// split feature set
+	return strings.Split(splitVersion[1], ","), nil
+}
+
+// GetCompatibleVersions returns a descending ordered set of compatible IBC
+// versions for the caller chain's connection end. The latest supported
+// version should be first element and the set should descend to the oldest
+// supported version.
 func GetCompatibleVersions() []string {
 	return []string{DefaultIBCVersion}
 }
 
-// IsSupportedVersion returns true if the version provided is supported.
-func IsSupportedVersion(version string, supportedVersions []string) bool {
-	for _, supportedVer := range supportedVersions {
-		if supportedVer == version {
-			return true
+// FindSupportedVersion returns the version with a matching version number
+// if it exists. The returned boolean is true if the version is found and
+// false otherwise.
+func FindSupportedVersion(version string, supportedVersions []string) (string, bool) {
+	versionNumber, err := GetVersionNumber(version)
+	if err != nil {
+		return "", false
+	}
+
+	for _, supportedVersion := range supportedVersions {
+		supportedVersionNumber, err := GetVersionNumber(supportedVersion)
+		if err != nil {
+			continue
+		}
+
+		if supportedVersionNumber == versionNumber {
+			return supportedVersion, true
 		}
 	}
-	return false
+	return "", false
 }
 
-// PickVersion iterates over the descending ordered set of compatible IBC versions
-// and selects the first version that is supported by the counterparty.
+// PickVersion iterates over the descending ordered set of compatible IBC
+// versions and selects the first version that is supported by the counterparty.
 func PickVersion(counterpartyVersions []string) (string, error) {
 	versions := GetCompatibleVersions()
 
 	for _, ver := range versions {
-		if IsSupportedVersion(ver, counterpartyVersions) {
-			return ver, nil
+		if counterpartyVer, found := FindSupportedVersion(ver, counterpartyVersions); found {
```

Comment:
So the preference order of the versions is the order in which the initiating chain passed in correct (in descending priority order)

Reply:
* OpenInit: partyA sets a set of versions
* OpenTry: partyB iterates over it's own supported versions (first for loop above), for each version it descends over, it checks if partyA supports this version, if so, it selects this version and the intersection of the feature set for these versions.

I guess another issue here is whether to select a version with a nil overlapping feature set or select a later version with a non nil feature set. The current code reflects the former. I think this is fine if we follow that the version number specifies a core set of features that are expected to be implemented, this would be appropriate default behaviour. A party could always set their preferences to have their versions with the most supported features first.

Reply:
Yes, this is correct, the initiating chain will then have a chance to confirm the choice in `OpenAck`.